src/Contacts.ts

```diff
@@ -0,0 +1,203 @@
+  }
+}
+
+export class AllowList {
+  entries: Map<string, AllowListPermissionType>
+  static _identifier: string
+
+  constructor() {
+    this.entries = new Map<string, AllowListPermissionType>()
+  }
+
+  allow(address: string) {
+    const entry = AllowListEntry.fromAddress(address, 'allow')
+    this.entries.set(entry.key, 'allow')
+    return entry
+  }
+
+  block(address: string) {
+    const entry = AllowListEntry.fromAddress(address, 'block')
+    this.entries.set(entry.key, 'block')
+    return entry
+  }
+
+  state(address: string) {
+    const entry = AllowListEntry.fromAddress(address)
+    return this.entries.get(entry.key) ?? 'unknown'
+  }
+
+  static async getIdentifier(client: Client): Promise<string> {
+    if (!this._identifier) {
+      const { identifier } =
+        await client.keystore.getPrivatePreferencesTopicIdentifier()
+      this._identifier = identifier
+    }
+    return this._identifier
+  }
+
+  static async load(client: Client): Promise<AllowList> {
+    const allowList = new AllowList()
+    const identifier = await this.getIdentifier(client)
+    const envelopes = (
+      await client.listEnvelopes(
+        buildUserPrivatePreferencesTopic(identifier),
+        async ({ message }: Envelope) => {
+          if (message) {
+            const result = await client.keystore.selfDecrypt({
+              requests: [{ payload: message }],
+            })
+            const payload = result.responses[0].result?.decrypted
+            if (payload) {
+              return privatePreferences.PrivatePreferencesAction.decode(payload)
+            }
+          }
+          return undefined
+        }
+      )
+    ).filter(
+      (envelope) => envelope !== undefined
+    ) as privatePreferences.PrivatePreferencesAction[]
+
+    envelopes.forEach((envelope) => {
+      envelope.allow?.walletAddresses.forEach((address) => {
+        allowList.allow(address)
+      })
+      envelope.block?.walletAddresses.forEach((address) => {
+        allowList.block(address)
+      })
+    })
+
+    return allowList
+  }
+
+  static async publish(entries: AllowListEntry[], client: Client) {
+    const identifier = await this.getIdentifier(client)
+
+    // TODO: preserve order
+    const rawEnvelopes = await Promise.all(
+      entries.map(async (entry) => {
+        if (entry.entryType === 'address') {
+          const action: privatePreferences.PrivatePreferencesAction = {
+            allow:
+              entry.permissionType === 'allow'
+                ? {
+                    walletAddresses: [entry.value],
+                  }
+                : undefined,
+            block:
+              entry.permissionType === 'block'
+                ? {
+                    walletAddresses: [entry.value],
+                  }
+                : undefined,
+          }
+          const payload =
+            privatePreferences.PrivatePreferencesAction.encode(action).finish()
+          const result = await client.keystore.selfEncrypt({
```

Comment:
Less important for encryption than decryption, but would be nice to take advantage of the batch capabilities of the API and do them all at once. At least in the context of the Snap, there's some overhead to each request.